Illegal access exception with macros (JRuby 9.1.15.0 & Java 9)

When you run a project with a simple macro (I have taken the java-extension-example project from the maven examples) with JRuby 9.1.15.0 (latest) and Java 9 you get an Exception like this:

> [ERROR] Failed to execute goal org.asciidoctor:asciidoctor-maven-plugin:1.5.6:process-asciidoc (asciidoctor-with-extension-example) on project asciidoctor-with-extension-example: Execution asciidoctor-with-extension-example of goal org.asciidoctor:asciidoctor-maven-plugin:1.5.6:process-asciidoc failed: org.jruby.exceptions.RaiseException: (TypeError) illegal access on 'process': class org.jruby.javasupport.JavaMethod cannot access a member of class org.asciidoctor.extension.MacroProcessor with modifiers "protected abstract" -> [Help 1]

I did not investigate it yet, but I posted the issue here because I think the problem comes from AsciidoctorJ and the newer versions of JRuby and the JVM and not from the specific maven example.

The issue does not appear with:

* Java 8 & JRuby 9.1.15.0
* Java 8 & JRuby 9.0.5.0
* Java 9 & JRuby 9.0.5.0

Tested on macOS.

In case it helps, the issue appears with jRuby 9.1.14.0 which is the first marked as supporting Java9 (http://jruby.org/news).
Previous versions that work fine for me: Maven 3.5.2, Java 9.0.4, Windows7.

```
Caused by: org.jruby.exceptions.RaiseException: (TypeError) illegal access on 'process': class org.jruby.javasupport.JavaMethod cannot access a member of class org.asciidocto
r.extension.MacroProcessor with modifiers "protected abstract"                                                                                                                
    at org.jruby.RubyMethod.call (org/jruby/RubyMethod.java:127)                                                                                                              
    at RUBY.block in sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:642)                                                               
    at org.jruby.RubyString.gsub (org/jruby/RubyString.java:2541)                                                                                                             
    at RUBY.block in sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:615)                                                               
    at org.jruby.RubyArray.each (org/jruby/RubyArray.java:1734)                                                                                                               
    at RUBY.sub_macros (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:614)                                                                        
    at RUBY.block in apply_subs (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:126)                                                               
    at org.jruby.RubyArray.each (org/jruby/RubyArray.java:1734)                                                                                                               
    at RUBY.apply_subs (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/substitutors.rb:115)                                                                        
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/block.rb:112)                                                                                  
    at RUBY.paragraph (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/html5.rb:719)                                                                      
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/base.rb:34)                                                                          
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:70)                                                                          
    at RUBY.block in content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:79)                                                                 
    at org.jruby.RubyArray.map (org/jruby/RubyArray.java:2486)                                                                                                                
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/abstract_block.rb:79)                                                                          
    at RUBY.content (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/document.rb:1120)                                                                              
    at RUBY.document (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/html5.rb:186)                                                                       
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/converter/base.rb:34)                                                                          
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor/document.rb:1060)                                                                              
    at RUBY.convert (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1502)                                                                                       
    at RUBY.block in convert_file (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575)                                                                         
    at org.jruby.RubyIO.open (org/jruby/RubyIO.java:1156)                                                                                                                     
    at RUBY.convert_file (uri:classloader:/gems/asciidoctor-1.5.6.1/lib/asciidoctor.rb:1575)                                                                                  
    at RUBY.convertFile (<script>:75)                                                                                                                                         
,,,

--- a/asciidoctorj-core/src/main/java/org/asciidoctor/extension/MacroProcessor.java
+++ b/asciidoctorj-core/src/main/java/org/asciidoctor/extension/MacroProcessor.java
@@ -30,6 +30,6 @@ public abstract class MacroProcessor extends Processor {
         return new HashMap<Object, Object>();
     }
     
-    protected abstract Object process(AbstractBlock parent, String target, Map<String, Object> attributes);
+    public abstract Object process(AbstractBlock parent, String target, Map<String, Object> attributes);
     
 }

--- a/asciidoctorj-core/src/test/java/org/asciidoctor/extension/ManpageMacro.java
+++ b/asciidoctorj-core/src/test/java/org/asciidoctor/extension/ManpageMacro.java
@@ -12,7 +12,7 @@ public class ManpageMacro extends InlineMacroProcessor {
     }
 
     @Override
-    protected String process(AbstractBlock parent, String target,
+    public String process(AbstractBlock parent, String target,
             Map<String, Object> attributes) {
         Map<String, Object> options = new HashMap<String, Object>();
         options.put("type", ":link");

--- a/asciidoctorj-core/src/test/java/org/asciidoctor/extension/WhenRubyExtensionIsRegistered.java
+++ b/asciidoctorj-core/src/test/java/org/asciidoctor/extension/WhenRubyExtensionIsRegistered.java
@@ -25,7 +25,7 @@ public class WhenRubyExtensionIsRegistered {
     public void ruby_extension_should_be_registered() {
         
         RubyExtensionRegistry rubyExtensionRegistry = this.asciidoctor.rubyExtensionRegistry();
-        rubyExtensionRegistry.loadClass(Class.class.getResourceAsStream("/YellRubyBlock.rb")).block("rubyyell", "YellRubyBlock");
+        rubyExtensionRegistry.loadClass(getClass().getResourceAsStream("/YellRubyBlock.rb")).block("rubyyell", "YellRubyBlock");
 
         String content = asciidoctor.renderFile(
                 classpath.getResource("sample-with-ruby-yell-block.ad"),
@@ -42,7 +42,7 @@ public class WhenRubyExtensionIsRegistered {
     public void ruby_extension_should_be_unregistered() {
 
         ExtensionGroup extensionGroup = this.asciidoctor.createGroup()
-            .loadRubyClass(Class.class.getResourceAsStream("/YellRubyBlock.rb"))
+            .loadRubyClass(getClass().getResourceAsStream("/YellRubyBlock.rb"))
             .rubyBlock("rubyyell", "YellRubyBlock");
 
         {

--- a/asciidoctorj-core/src/test/java/org/asciidoctor/internal/WhenClassloaderIsRequired.java
+++ b/asciidoctorj-core/src/test/java/org/asciidoctor/internal/WhenClassloaderIsRequired.java
@@ -3,6 +3,7 @@ package org.asciidoctor.internal;
 import static org.hamcrest.CoreMatchers.is;
 import static org.junit.Assert.assertThat;
 import org.asciidoctor.Asciidoctor;
+import org.junit.Ignore;
 import org.junit.Test;
 
  class AsciiDoctorJClassloaderTestRunnable implements Runnable {
@@ -46,6 +47,7 @@ public class WhenClassloaderIsRequired {
         inside the runnable to verify we have created the same error as happens inside sbt.
      */
     @Test
+    @Ignore("Behavior changed in JRuby 9000. It no longer only uses the TCCL by default")
     public void contentsOfJRubyCompleteShouldFailToLoadWithoutPassingClassloader() throws Exception{
         ClassLoader currentclassloader =  this.getClass().getClassLoader();
         ClassLoader rootclassloader =  currentclassloader.getParent();

